src/index/silentpaymentindex.cpp

```diff
@@ -0,0 +1,196 @@
+#include <index/silentpaymentindex.h>
+
+#include <chainparams.h>
+#include <coins.h>
+#include <index/disktxpos.h>
+#include <node/blockstorage.h>
+#include <pubkey.h>
+#include <undo.h>
+#include <util/system.h>
+#include <validation.h>
+
+using node::ReadBlockFromDisk;
+using node::UndoReadFromDisk;
+
+constexpr uint8_t DB_SILENTPAYMENTINDEX{'s'};
+
+std::unique_ptr<SilentPaymentIndex> g_silentpaymentindex;
+
+/** Access to the silent payment index database (indexes/silentpaymentindex/) */
+class SilentPaymentIndex::DB : public BaseIndex::DB
+{
+public:
+    explicit DB(size_t n_cache_size, bool f_memory = false, bool f_wipe = false);
+
+    bool WriteSilentPayments(const std::vector<std::pair<uint256, uint256>>& items);
+};
+
+SilentPaymentIndex::DB::DB(size_t n_cache_size, bool f_memory, bool f_wipe) :
+    BaseIndex::DB(gArgs.GetDataDirNet() / "indexes" / "silentpaymentindex", n_cache_size, f_memory, f_wipe)
+{}
+
+bool SilentPaymentIndex::DB::WriteSilentPayments(const std::vector<std::pair<uint256, uint256>>& items)
+{
+    CDBBatch batch(*this);
+    for (const auto& tuple : items) {
+        batch.Write(std::make_pair(DB_SILENTPAYMENTINDEX, tuple.first), tuple.second);
+    }
+    return WriteBatch(batch);
+}
+
+SilentPaymentIndex::SilentPaymentIndex(size_t n_cache_size, bool f_memory, bool f_wipe)
+    : m_db(std::make_unique<SilentPaymentIndex::DB>(n_cache_size, f_memory, f_wipe))
+{}
+
+SilentPaymentIndex::~SilentPaymentIndex() {}
+
+bool SilentPaymentIndex::ExtractPubkeyFromInput(const Coin& prevCoin, const CTxIn& txin,  XOnlyPubKey& senderPubKey)
+{
+    // scriptPubKey being spent by this input
+    CScript scriptPubKey = prevCoin.out.scriptPubKey;
+
+    if (scriptPubKey.IsPayToWitnessScriptHash()) {
+        return false;
+    }
+
+    // Vector of parsed pubkeys and hashes
+    std::vector<std::vector<unsigned char>> solutions;
+
+    TxoutType whichType = Solver(scriptPubKey, solutions);
+
+    if (whichType == TxoutType::NONSTANDARD ||
+    whichType == TxoutType::MULTISIG ||
+    whichType == TxoutType::WITNESS_UNKNOWN ) {
+        return false;
+    }
+
+    const CScript scriptSig = txin.scriptSig;
+    const CScriptWitness scriptWitness = txin.scriptWitness;
+
+    assert(senderPubKey.IsNull());
+
+    // TODO: Condition not tested
+    if (whichType == TxoutType::PUBKEY) {
+
+        CPubKey pubkey = CPubKey(solutions[0]);
+        assert(pubkey.IsFullyValid());
+        senderPubKey = XOnlyPubKey(pubkey);
+    }
+
+    else if (whichType == TxoutType::PUBKEYHASH) {
+
+        int sigSize = static_cast<int>(scriptSig[0]);
+        int pubKeySize = static_cast<int>(scriptSig[sigSize + 1]);
+        auto serializedPubKey = std::vector<unsigned char>(scriptSig.begin() + sigSize + 2, scriptSig.end());
+        assert(serializedPubKey.size() == pubKeySize);
```

Comment:
`warning:` comparison of integer expressions of different signedness: ‘std::vector<unsigned char>::size_type’ {aka ‘long unsigned int’} and ‘int’ [-Wsign-compare]

Reply:
Done in https://github.com/bitcoin/bitcoin/pull/24897/commits/6120b53da4b1a7dc067790a1c00ac744a66f2f71. Thanks.